### ISSUE TITLE
Support of PHP versions?

### ISSUE THREAD

**Issue by stephanvierkant** [liip/LiipImagineBundle#1438]
I see this library still supports PHP 7.1, which is End of Life (EOL) for a long time. Even 7.3 is EOL as of today.

What's the policy about PHP support? The only [supported versions](https://www.php.net/supported-versions.php) are 7.4, 8.0 and 8.1

**Comment by dbu** [comment]
as long as it does not cause overhead, we keep supporting legacy PHP versions. when that becomes an issue, we would drop support for old versions in a minor version change.

that said, we are currently working on a new major version, and i think we will drop some EOL PHP versions to be able to use new PHP language features.

out of curiosity: what is your reason for asking the question?

**Comment by stephanvierkant** [comment]
When working on open source projects, I see many of them still support legacy PHP versions. That makes it harder (or at least less fun) to contribute, because you can't use newer features. That's why I looked at the composer.lock of some of the projects I'm working on to see what projects still support EOL-versions and/or don't support Symfony 6, to see if I can help.

In my opinion, following PHP's policy (drop support when it's EOL) can help projects to stay maintainable. But to be honest, I didn't see the 3.x branch already dropped support. I'll see if I can help with the [open issues](https://github.com/liip/LiipImagineBundle/issues?q=is%3Aopen+is%3Aissue+milestone%3A3.0.0)!

**Comment by dbu** [comment]
great, glad if you want to contribute something. #1415 would be all about using new language features and dropping old PHP versions ;-)

**Comment by pvisback** [comment]
Hello guys. Is the support of PHP 8.1 (without deprecation warnings) scheduled in a near future, or should I mute these warnings for now ? I'm having warnings in MetadataBag, about the need to use ReturnTypeWillChange attribute in order to have consistent return types. Thanks

**Comment by dbu** [comment]
the library is tested and works fine with PHP 8.1. the deprecation warnings are just warnings, they don't mean anything is broken. they will have to be solved before we can support PHP 9, so contributions to solve these warnings are appreciated. until they land, you can safely mute warnings for this library (until you plan to upgrade to PHP 9, that is)

**Comment by tacman** [comment]
Symfony rocks, but one area I've never quite figured out is getting rid of deprecation warnings in dependencies.   In particular, the warning about ReturnTypeWillChange seems to show up a lot.

Is there a way to give preferential treatment to the latest release of PHP (and Symfony)?  That is, use a docblock or attribute to quiet the deprecation?  

Since version 2 of this bundle supports PHP 7.4, Version 3 could require PHP 8.0+, which would be consistent with Symfony 6 requirements.

I think the OP was stating the same thing -- it's much more exciting to dive into a bundle with the latest and greatest features.  I helped with getting a few bundles working with Symfony 6, but I kept getting stuck on how to do things that required BC support with older versions.  So if V3 were 8.0+ and Symfony 6+, and V2 continued to support older versions, then we could easily use the new language features, some of which really leverage static analysis, finding bugs without even writing tests.

**Comment by dbu** [comment]
yeah, BC for old symfony versions can be quite a hassle, been there and done that too :-)

i actually can get on board with that plan for version 3, to be PHP 8 and SF 6 only. somebody started with return types in #1439 but there are a few things that probably need to be cleaned up.

are you motivated to work on the 3.x branch to bring full static typing where possible? if you do, please do small-ish PRs that either do one thing or only touch one namespace to keep the reviews manageable and be able to merge them in timely fashion. the github issue for the topic of static typing is #1415 btw, and there is a couple of other ideas in the 3.0 milestone: https://github.com/liip/LiipImagineBundle/milestone/18
